### PR TITLE
Sign out and lock functionality

### LINKED ISSUE #81: Enable/disable biometric lock (settings)
## Acceptance criteria

- user should be able to enable it from settings
  - if hasn't already been granted, will prompt user for access to biometrics
- user should be able to disable it from settings
  - which means they have to FxA Authenticate if auto lock setting enabled...

## Visual Design
**Invision**
https://mozilla.invisionapp.com/share/XCEV238ZD#/267560609_Settings_-_Touch_ID_Enabled

**Zeplin**
https://zpl.io/VKNMd5W

### LINKED ISSUE #171: Sign-out to lock (settings)
When I want to lock my account manually, I need an option to 'sign out' with the ability to easily unlock the app again later.

## Acceptance criteria

- Should be able to lock saved logins
- Should be able to unlock with biometrics
- If biometrics access not set, should be able to sign back in with FxA "master password"

## Visual Design
**InVision**
Sign Out (settings): https://mozilla.invisionapp.com/share/XCEV238ZD#/267560609_Settings_-_Touch_ID_Enabled
Sign In - Returning User (screen after signing out): https://mozilla.invisionapp.com/share/XCEV238ZD#/267560535_Welcome_-_Returning_User

**Zeplin**
Sign Out (settings): https://zpl.io/VKNMd5W
Sign In - Returning User (screen after signing out): https://zpl.io/V0ZyORR

### PR DESCRIPTION
**Pull request by sashei**
resolves #171 
resolves #81

### CODE REVIEW

**Comment by sashei** [comment]
@devinreams yeah, the double entries has to do with the test data so I wasn't concerned about it.

I'll look into the faceID thing though, thanks for catching that.

**Comment by sashei** [comment]
To get the correct prompt to show up in the simulator, you have to select `Hardware` -> `Face ID` -> `Enrolled`.

I think there's an open question here about what to do when the user is *not* enrolled in biometrics at all, but I wonder if it's worth shaking out the edge cases in this particular PR or addressing them in #35. I can follow-up on this by putting together a list in that issue.

**Comment by devinreams** [comment]
>To get the correct prompt to show up in the simulator, you have to select Hardware -> Face ID -> Enrolled.

Aha! I see it now...

> I wonder if it's worth shaking out the edge cases in this particular PR or addressing them in #35

I'm good with keeping this light and getting it "feature complete", merged, and then follow up in the list in #35. 👍 This is going to make for a fun test matrix..

**Comment by devinreams** [comment]
@sashei in case you didn't see it, the last two merges to `master` caused some conflicts here ⚠️

File: lockbox-ios/Presenter/AutoLockSettingsPresenter.swift [hunk 1]
```diff
@@ -55,7 +54,7 @@ class AutoLockSettingPresenter {
     }
 
     func onViewReady() {
-        let driver = self.userDefaults.rx.observe(String.self, SettingKey.autoLock.rawValue)
+        let driver = self.userDefaults.rx.observe(String.self, SettingKey.autoLockTime.rawValue)
```
Comment (jhugman) [inline lockbox-ios/Presenter/AutoLockSettingsPresenter.swift]: Use

```swift
userDefaults.on(setting: .autoLockTime, type: String.self)
```

or better, put this in the `UserDefaults` extension.

File: lockbox-ios/Presenter/SettingListPresenter.swift [hunk 2]
```diff
@@ -55,29 +49,35 @@ class SettingsPresenter {
...
     func onViewReady() {
-        let biometricObserver = self.userDefaults.rx.observe(Bool.self, SettingKey.biometricLogin.rawValue)
-        let autoLockObserver = self.userDefaults.rx.observe(String.self, SettingKey.autoLock.rawValue).filterNil()
+        let autoLockObserver = self.userDefaults.rx.observe(String.self, SettingKey.autoLockTime.rawValue).filterNil()
 
-        let settingsConfigDriver = Observable.combineLatest(biometricObserver, autoLockObserver)
-                .map { (latest: (Bool?, String)) -> [SettingSectionModel] in
+        let settingsConfigDriver = Observable.combineLatest(self.userDefaults.onBiometricsEnabled, autoLockObserver)
+                .map { (latest: (Bool, String)) -> [SettingSectionModel] in
                     let autoLock = AutoLockSetting(rawValue: latest.1) ?? AutoLockSetting.FiveMinutes
```
Comment (jhugman) [inline lockbox-ios/Presenter/SettingListPresenter.swift]: Use:

```swift
let autoLock = AutoLockSetting(rawValue: latest.1) ?? Constants.setting.defaultAutoLockTimeout
```

File: lockbox-ios/Presenter/WelcomePresenter.swift [hunk 3]
```diff
@@ -5,24 +5,76 @@
...
+    var biometricAuthenticationPromptHidden: AnyObserver<Bool> { get }
+    var biometricSignInText: AnyObserver<String?> { get }
+    var biometricImageName: AnyObserver<String> { get }
+    var fxAButtonTopSpace: AnyObserver<CGFloat> { get }
+}
+
+struct LockedEnabled {
+    let locked: Bool
+    let enabled: Bool
+}
```
Comment (jhugman) [inline lockbox-ios/Presenter/WelcomePresenter.swift]: Nit: document what Locked and Enabled mean.

File: lockbox-ios/Common/Extensions/UserDefaults+.swift [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import RxSwift
+import RxCocoa
+
+extension UserDefaults {
+    var onLock: Observable<Bool> {
+        return self.rx.observe(Bool.self, SettingKey.locked.rawValue).filterNil()
+    }
+
+    var onBiometricsEnabled: Observable<Bool> {
+        return self.rx.observe(Bool.self, SettingKey.biometricLogin.rawValue).filterNil()
```
[the remark this answers is not in the document]
Comment (sashei) [reply lockbox-ios/Common/Extensions/UserDefaults+.swift]: I like the refactor.

All of the NS rx extensions are a part of `RxCocoa`, provided as part of `RxSwift` :D

**Comment by devinreams** [comment]
@sashei I'm not sure how I did it but I'm got to a blank Settings screen view _except for_ the Sign Out button. 

screen shot 2018-04-12 at 10 29 12 am

I went back and hardware-enabled Face ID, deleted the app and refreshed everything fully (I believe) and now it's all back.

I share in case that's unexpected and helpful.

**Comment by sashei** [comment]
@devinreams ok, I think that's to be expected... the settings list relies on all of the defaults for the user settings being set. they are only set on the first run of the app (presumably we don't want to overwrite someone's settings) and when unlinking your account, so when installing this update people will need to do a full delete of the app rather than installing over a previous version.
